core/tx_pool.go

```diff
@@ -839,9 +839,19 @@ func (pool *TxPool) validateStakingTx(tx *staking.StakingTransaction) error {
 			return err
 		}
 		pendingEpoch := pool.pendingEpoch()
-		_, _, _, err = VerifyAndDelegateFromMsg(
+		_, delegateAmt, _, err := VerifyAndDelegateFromMsg(
 			pool.currentState, pendingEpoch, stkMsg, delegations, pool.chainconfig.IsRedelegation(pendingEpoch))
-		return err
+		if err != nil {
+			return err
+		}
+		// We need to deduct gas price and verify balance since txn.Cost() is not accurate for delegate
+		// staking transaction because of re-delegation.
+		gasAmt := new(big.Int).Mul(tx.GasPrice(), new(big.Int).SetUint64(tx.GasLimit()))
+		totalAmt := new(big.Int).Add(delegateAmt, gasAmt)
+		if bal := pool.currentState.GetBalance(from); bal.Cmp(totalAmt) < 0 {
```

Comment:
very nice catch!. Please also check on other staking txns to make sure there is no verification issue like this where the gas cost is not counted.

Reply:
Other staking transactions are good since their costs are calculated correctly and checked with `stakingTransaction.Cost()`. For delegation, the cost need more comprehensive logic handling along with statedb because of redelegation. So I think other staking transactions are just fine and gas fee is calculated.